iroh/src/magicsock/relay_actor.rs

```diff
@@ -152,265 +224,455 @@ impl ActiveRelayActor {
             #[cfg(any(test, feature = "test-utils"))]
             insecure_skip_cert_verify,
         } = opts;
-        let mut builder = relay::client::ClientBuilder::new(url)
+        let mut builder = relay::client::ClientBuilder::new(url, secret_key, dns_resolver)
             .address_family_selector(move || prefer_ipv6.load(Ordering::Relaxed));
         if let Some(proxy_url) = proxy_url {
             builder = builder.proxy_url(proxy_url);
         }
         #[cfg(any(test, feature = "test-utils"))]
         let builder = builder.insecure_skip_cert_verify(insecure_skip_cert_verify);
-        builder.build(secret_key, dns_resolver)
+        builder
     }
 
-    async fn run(mut self, mut inbox: mpsc::Receiver<ActiveRelayMessage>) -> anyhow::Result<()> {
+    /// The main actor run loop.
+    ///
+    /// Primarily switches between the dialing and connected states.
+    async fn run(mut self) -> anyhow::Result<()> {
         inc!(MagicsockMetrics, num_relay_conns_added);
-        debug!("initial dial {}", self.url);
-        self.relay_client
-            .connect()
-            .await
-            .context("initial connection")?;
 
-        // When this future has an inner, it is a future which is currently sending
-        // something to the relay server.  Nothing else can be sent to the relay server at
-        // the same time.
-        let mut relay_send_fut = std::pin::pin!(MaybeFuture::none());
+        loop {
+            let Some(client) = self.run_dialing().instrument(info_span!("dialing")).await else {
+                break;
+            };
+            match self
+                .run_connected(client)
```

Comment:
I don't know if it makes this code more complex, but I am wondering if it would be better to model the state like this

```rust
enum ActiveRelayConnector {
  Dialing {
    // ..
  },
  Connected {
    // .. 
  }
}
```

Reply:
Currently the state in the `ActiveRelayActor` itself is used in all 3 states the actor can be in.  The `ConnectedRelayState` is only shared between sending and connected states.  I don't feel like trying to put this state in the actor itself will help much, you'd still have `run_sending` implemented in the disconnected state, but this is already alleviated by it needing the `ConnectedRelayState`.  What might work is create `ConnectedRelay` and `DisconnectedRelay` and impl the methods on there.  They could take a ref to the data in the `ActiveRelayActor` probably.

Reply:
https://github.com/n0-computer/iroh/pull/3095